ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/MiscHelpers.java

```diff
@@ -177,6 +182,52 @@ private IntList computeCommitteeShuffle(
         .subList(fromIndex, toIndex);
   }
 
+  public List<UInt64> computeSubscribedSubnets(final UInt256 nodeId, final UInt64 epoch) {
+    return IntStream.range(0, specConfig.getNetworkingConfig().getSubnetsPerNode())
+        .mapToObj(index -> computeSubscribedSubnet(nodeId, epoch, index))
+        .collect(Collectors.toList());
+  }
+
+  public UInt64 computeSubscribedSubnet(final UInt256 nodeId, final UInt64 epoch, final int index) {
+
+    final int nodeIdPrefix =
+        nodeId
+            .shiftRight(
+                NetworkConstants.NODE_ID_BITS
+                    - specConfig.getNetworkingConfig().getAttestationSubnetPrefixBits())
+            .intValue();
+
+    final UInt64 nodeOffset =
+        UInt64.valueOf(
+            nodeId.mod(specConfig.getNetworkingConfig().getEpochsPerSubnetSubscription()).toLong());
+
+    final Bytes32 permutationSeed =
+        Hash.sha256(
+            uint64ToBytes(
+                epoch
+                    .plus(nodeOffset)
+                    .dividedBy(specConfig.getNetworkingConfig().getEpochsPerSubnetSubscription())));
+
+    final int permutedPrefix =
+        computeShuffledIndex(
+            nodeIdPrefix,
+            1 << specConfig.getNetworkingConfig().getAttestationSubnetPrefixBits(),
+            permutationSeed);
+
+    return UInt64.valueOf(
+        permutedPrefix + index % specConfig.getNetworkingConfig().getAttestationSubnetCount());
+  }
+
+  public UInt64 calculateNodeSubnetUnsubscriptionSlot(final UInt64 currentSlot) {
```

Comment:
Ok, as I finally got my answers and clear on this, how should we calculate this slot:
It should be the last slot of the epoch prior to change in subnetIds computation result
So this method should have logic something like this:
```
(nodeId, currentSlot)
final UInt64 nodeOffset =
        UInt64.valueOf(
            nodeId.mod(specConfig.getNetworkingConfig().getEpochsPerSubnetSubscription()).toLong());
currentEpoch = f(currentSlot)
nextPeriodEpoch = ((currentEpoch + specConfig.getNetworkingConfig().getEpochsPerSubnetSubscription()) % specConfig.getNetworkingConfig().getEpochsPerSubnetSubscription()) - nodeOffset
nextPeriodEpochFirstSlot = f(nextPeriodEpoch)
return nextPeriodEpochFirstSlot - 1
```
We need some testing for this, my first idea is iterate over 1000 sequential epochs with several random nodeIds and the results of compute and this function will match. Any extra tests on this will be good too

Reply:
Done. I implemented the slot calculation slightly differently and added a test. will see if I can test it further. Let me know what you think about it now.